### PR TITLE
Generalize TransferCoin to TransferObject

### LINKED ISSUE #2397: [Sui Framework] Generalize TransferCoin to TransferObject
- Generalize TransferCoin to TransferObject, predicated on the type having store
- Be sure to used BinaryIndexedView::abilities. It is not enough to look at the declared abilities

See also #2173

### PR DESCRIPTION
**Pull request by tnowacki**
- Fixes #2397
- TransferObject works over any type with store
- Added a field to MoveObject in order to store this flag
- Adding a field to MoveObject is a tad invasive, and increases storage for literally every Move object. But I think it is better than the alternative, which would involve spinning up the VM for every `TransferCoin` (now `PublicTransferObject`) request. And I think we really want these to be as fast as possible. 
- I went with the name `PublicTransferObject` to indicate that not all objects can be transferred, but maybe that name doesn't make it clear... 

## Tests

- I will add adapter tests in a following PR
- Not sure what else to test
- ~~it looks like I broke a test and am a bit unsure how~~

### CODE REVIEW

File: sdk/typescript/src/types/transactions.ts [hunk 1]
```diff
@@ -4,15 +4,15 @@
 import { ObjectOwner, SuiAddress, TransactionDigest } from './common';
 import { SuiMovePackage, SuiObject, SuiObjectRef } from './objects';
 
-export type TransferCoin = {
+export type PublicTransferObject = {
```
Comment (gegaowp) [inline sdk/typescript/src/types/transactions.ts:7]: Thanks for making the change Todd!

cc @666lcz looks like we can leverage this for NFT as well. we will need to change the `signer & provider` accordingly, probably also `sui-objects/Coin.ts` before migrating it to SDK or a shared lib, right?

File: crates/sui-types/src/object.rs [hunk 2]
```diff
@@ -34,6 +33,9 @@ pub const OBJECT_START_VERSION: SequenceNumber = SequenceNumber::from_u64(1);
 #[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
 pub struct MoveObject {
     pub type_: StructTag,
+    /// Determines if it is usable with the PublicTransferObject
+    /// Derived from the type_
+    has_public_transfer: bool,
```
Comment (666lcz) [inline crates/sui-types/src/object.rs:38]: Can we also add this field to [SuiMoveObject](https://github.com/MystenLabs/sui/blob/764f4fd9d2dff37be0345080ada8d103e87de23b/crates/sui-core/src/gateway_types.rs#L323)? `MoveObject` will get translated into `SuiMoveObject` in the RPC layer to make sure RPC has a stable API. Having `has_public_transfer` exposed to the RPC client will enable the client(e.g. wallet) to tell if the object is transferrable

cc @patrickkuo

File: sdk/typescript/src/types/transactions.ts (hunk 1 above)
Comment (666lcz) [reply sdk/typescript/src/types/transactions.ts:7]: yes, we also need to update https://github.com/MystenLabs/sui/blob/8a3d0789caecb5751036a2b308a9c64f600c7cbe/sdk/typescript/src/signers/signer-with-provider.ts#L71 , but I can help make this change

**Review by lxfind — APPROVED** [review]
Approve so it can go in asap to avoid more merge conflicts.
Can you summarize the list of namings introduced in this PR and post in sui-eng?

**Comment by Clay-Mysten** [comment]
Let's definitely make sure we update docs accordingly. Please let me know how I can help.

**Comment by Clay-Mysten** [comment]
> Let's definitely make sure we update docs accordingly. Please let me know how I can help.

I see the Markdown file updates and so assume we accounted for doc updates. I also did some grepping and see no references to TransferCoin remaining.

Please let me know if I can help ensure this change goes smoothly, such as an announcement in:
https://docs.sui.io/learn#see-whats-new
